### PR TITLE
Improve code coverage for NegoState class

### PR DESCRIPTION
**Pull request by caesar1995**
Contributes to #17099

Before the PR:

System.Net.Security | 6025 | 2721 | 8746 | 32749 | 68.8% |  55.2%
-- | -- | -- | -- | -- | --  | --
System.Net.Security.NegoState | 267 | 229 | 496 | 843 | **_53.8%_** | **_46.5%_**

After PR:

System.Net.Security | 6311 | 2432 | 8743 | 32744 | 72.1% | 58.1%
-- | -- | -- | -- | -- | --  | --
System.Net.Security.NegoState | 392 | 101 | 493 | 838 | _**79.5%**_ | _**71.7%**_

Summary:
- Clean a redundant check in `NegoState`.
- Add more test cases for corner cases check in `NegoState` class.

### CODE REVIEW

**Comment by caesar1995** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build
@dotnet-bot test Outerloop NETFX x86 Debug Build
@dotnet-bot test Outerloop UWP CoreCLR x64 Debug Build
@dotnet-bot test Outerloop OSX x64 Debug Build

File: src/System.Net.Security/src/System/Net/Security/NegoState.cs [hunk 1]
```diff
@@ -53,11 +53,6 @@ internal class NegoState
 
         internal NegoState(Stream innerStream, bool leaveStreamOpen)
         {
-            if (innerStream == null)
```
Comment (caesar1995) [inline src/System.Net.Security/src/System/Net/Security/NegoState.cs]: The only place to create a `NegoState` is [here](https://github.com/dotnet/corefx/blob/a44c93063d24e29971891bfc2cf8b0aa8771eae9/src/System.Net.Security/src/System/Net/Security/NegotiateStream.cs#L44). Before passing the `innerStream` to `NegoState`'s ctor, the `NegotiateStream`'s base class `AuthenticatedStream` will check the stream for null value. So the check here is not needed.

**Comment by caesar1995** [comment]
The CI is stuck, but I think the PR is ready for review.

File: src/System.Net.Security/src/System/Net/Security/NegoState.cs (hunk 1 above)
Comment (stephentoub) [reply src/System.Net.Security/src/System/Net/Security/NegoState.cs]: In such cases we often then assert:
```C#
Debug.Assert(innerStream != null);
```

Comment (caesar1995) [reply src/System.Net.Security/src/System/Net/Security/NegoState.cs]: Good to know that! Will fix.

File: src/System.Net.Security/tests/FunctionalTests/NegotiateStreamStreamToStreamTest.cs [hunk 2]
```diff
@@ -90,6 +90,8 @@ public abstract class NegotiateStreamStreamToStreamTest
             {
                 Assert.False(client.IsAuthenticated);
                 Assert.False(server.IsAuthenticated);
+                Assert.Equal(false, client.IsMutuallyAuthenticated);
+                Assert.Equal(false, server.IsMutuallyAuthenticated);
```
Comment (davidsh) [inline src/System.Net.Security/tests/FunctionalTests/NegotiateStreamStreamToStreamTest.cs]: These 'Assert.Equal(false,' should be 'Assert.False('

Comment (davidsh) [reply src/System.Net.Security/tests/FunctionalTests/NegotiateStreamStreamToStreamTest.cs]: I don't understand why this test is named "_Success".  Nothing seems successful if all those properties are false.

What is this testing exactly? I think it is testing a failure condition where the stream ends up un-authenticated. So, perhaps the test should be named like:

"NegotiateStream_StreamToStream_Authentication_UseInvalidTargetName_UnauthenticatedStream"

Comment (caesar1995) [reply src/System.Net.Security/tests/FunctionalTests/NegotiateStreamStreamToStreamTest.cs]: > Nothing seems successful if all those properties are false.
> What is this testing exactly?

This test is asserting some properties values in NegotiateStream before and after authentication, to verify that the authentication with TargetName can succeed. This block of `Assert.False()` happens before calling into AuthenticateAsClientAsync/AuthenticateAsServerAsync, so those values are expected to be false. Therefore, the name to describe the purpose of this test method is correct.

Will fix the `Assert.Equal(false` problem.

File: src/System.Net.Security/tests/FunctionalTests/NegotiateStreamInvalidOperationTest.cs [hunk 3]
```diff
@@ -163,6 +169,200 @@ public class NegotiateStreamInvalidOperationTest
...
+            using (var server = new NegotiateStream(serverStream))
+            {
+                try
+                {
+                    // ProtectionLevel not match.
+                    await TestConfiguration.WhenAllOrAnyFailedWithTimeout(
+                        client.AuthenticateAsClientAsync((NetworkCredential)CredentialCache.DefaultCredentials,
+                            TargetName, ProtectionLevel.None, TokenImpersonationLevel.Identification),
+                        server.AuthenticateAsServerAsync((NetworkCredential)CredentialCache.DefaultCredentials,
+                            ProtectionLevel.Sign, TokenImpersonationLevel.Identification));
```
[the remark this answers is not in the document]
Comment (caesar1995) [reply src/System.Net.Security/tests/FunctionalTests/NegotiateStreamInvalidOperationTest.cs]: Fixed.

**Comment by caesar1995** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build
@dotnet-bot test Outerloop NETFX x86 Debug Build
@dotnet-bot test Outerloop UWP CoreCLR x64 Debug Build
@dotnet-bot test Outerloop OSX x64 Debug Build

**Pull request merged**